File: leetcode/intervals/57. Insert Interval/py/main.py
# Time O(N)
# Space O(1)
class Solution:
    def insert(self, intervals, newInterval):
      if not intervals: return [newInterval]

      if newInterval[1] < intervals[0][0]:
        return [newInterval] + intervals

      if newInterval[0] > intervals[-1][1]:
        return intervals + [newInterval]

      res = []
      isExtended = False

      for i in range(len(intervals)):
        if intervals[i][1] >= newInterval[0] and newInterval[1] >= intervals[i][0]:
          newInterval[0] = min(newInterval[0], intervals[i][0])
          newInterval[1] = max(newInterval[1], intervals[i][1])
          isExtended = True
        else:
          if not isExtended:
            if newInterval[1] < intervals[i][0]:
              res.append(newInterval)
              newInterval = intervals[i]
            elif intervals[i][1] < newInterval[0]:
              res.append(intervals[i])
          else:
            res.append(newInterval)
            newInterval = intervals[i]
      
      res.append(newInterval)
        
      return res

File: leetcode/intervals/57. Insert Interval/js/main.js
function insert(intervals, newInterval) {
  if (intervals.length === 0) return [newInterval];

  let result = [];
  let min = Number.MAX_VALUE;
  let max = -Number.MAX_VALUE;
  let i = 0;

  while (i < intervals.length) {
    if (overlap(intervals[i], newInterval)) {
      min = Math.min(min, intervals[i][0], newInterval[0]);
      max = Math.max(max, intervals[i][1], newInterval[1]);

      if (i + 1 === intervals.length) {
        result.push([min, max]);
      }
    } else {
      if (min === Number.MAX_VALUE && max === -Number.MAX_VALUE) {
        if (result.length > 0 && result[result.length - 1][1] < newInterval[0] && newInterval[1] < intervals[i][1]) {
          result.push(newInterval);
        }

        result.push(intervals[i]);
      } else {
        result.push([min, max]);
        result.push(intervals[i]);
        break;
      }
    }
    i++;
  }

  if (i < intervals.length) {
    return [...result, ...intervals.slice(i + 1)];
  }

  if (min === Number.MAX_VALUE && max === -Number.MAX_VALUE) {
    if (newInterval[0] > intervals[intervals.length - 1][0]) {
      result.push(newInterval);
    } else if (newInterval[1] < intervals[0][0]) {
      result.unshift(newInterval);
    }
  }
  return result;
}

function overlap([x, y], [u, z]) {
  return x <= z && u <= y;
}
